components/percona_telemetry/data_provider.cc

```diff
@@ -0,0 +1,317 @@
+   If there is replication and semisync_master or semisync_slave plugin is
+installed -> semi-sync replication For GR detection fool around GR plugin and
+performance_schema.replication_group_members table (see Orchestrator)
+2. product version with “…-pro” suffix for Pro Builds - SELECT VERSION();
+
+Should have:
+1. plugin information (list of active plugins) -
+    select plugin_name, plugin_status from information_schema.plugins;
+    select component_urn from mysql.component;
+
+Nice to have:
+1. MySQL uptime - SHOW GLOBAL STATUS LIKE 'Uptime';
+2. number of databases - SELECT COUNT(*) FROM information_schema.SCHEMATA WHERE
+SCHEMA_NAME NOT IN('mysql', 'information_schema', 'performance_schema', 'sys');
+3. size of databases - SELECT IFNULL(ROUND(SUM(data_length + index_length), 1),
+"0") size_MB FROM information_schema.tables WHERE table_schema NOT IN('mysql',
+'information_schema', 'performance_schema', 'sys');
+4. encryption methods applied (?)
+   encrypted tables count: SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE
+CREATE_OPTIONS LIKE '%ENCRYPTION%'; encrypted databases count: SELECT COUNT(*)
+FROM INFORMATION_SCHEMA.SCHEMATA WHERE DEFAULT_ENCRYPTION='YES'; encrypted
+tablespaces count: SELECT COUNT(*) FROM INFORMATION_SCHEMA.INNODB_TABLESPACES
+WHERE (flag & 8192) != 0;
+5. number of replication nodes, master/slave
+  source:
+  show replicas;
+  replica:
+  show replica status; -> Replica_IO_Running, Replica_SQL_Running
+6. storage engine used (MyRocks, InnoDB) - SELECT DISTINCT ENGINE FROM
+information_schema.tables WHERE table_schema NOT IN('mysql',
+'information_schema', 'performance_schema', 'sys'); 7.
+*/
+
+DataProvider::DataProvider(
+    SERVICE_TYPE(mysql_command_factory) & command_factory_service,
+    SERVICE_TYPE(mysql_command_options) & command_options_service,
+    SERVICE_TYPE(mysql_command_query) & command_query_service,
+    SERVICE_TYPE(mysql_command_query_result) & command_query_result_service,
+    SERVICE_TYPE(mysql_command_field_info) & command_field_info_service,
+    SERVICE_TYPE(mysql_command_thread) & command_thread_service, Logger &logger)
+    : command_factory_service_(command_factory_service),
+      command_options_service_(command_options_service),
+      command_query_service_(command_query_service),
+      command_query_result_service_(command_query_result_service),
+      command_field_info_service_(command_field_info_service),
+      command_thread_service_(command_thread_service),
+      logger_(logger) {}
+
+void DataProvider::thread_access_begin() { command_thread_service_.init(); }
+
+void DataProvider::thread_access_end() { command_thread_service_.end(); }
+
+bool DataProvider::do_query(const std::string &query, QueryResult *result) {
+  MYSQL_RES_H mysql_res = nullptr;
+  MYSQL_ROW_H row = nullptr;
+  uint64_t row_count = 0;
+  unsigned int num_column = 0;
+  std::string result_set;
+  MYSQL_H mysql_h = nullptr;
+  bool res = true;
+
+  if (command_factory_service_.init(&mysql_h)) {
```

Comment:
May be, use some RAII guard like std::unique_ptr with custom deleted for both mysql_h and mysql_res?

Reply:
I'm not sure if it is an overkill, but let's try this way... ;)